### PR TITLE
[12.0] [NEW] sale_require_po_doc: Added new module to check sale documentation.

### PR DESCRIPTION
**Pull request by Chandresh-SerpentCS**

### CODE REVIEW

**Comment by Chandresh-SerpentCS** [comment]
@dreispt Request for Review.

File: sale_require_po_doc/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+{
+    "name": "Sale Orders Require PO or Sales Documentation",
+    "version": "12.0.1.0.0",
+    "license": "AGPL-3",
+    "author": "Open Source Integrators, Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/sale-workflow",
+    "category": "Sales",
+    "depends": [
+        "sale_management", "purchase"
```
Comment (dreispt) [inline sale_require_po_doc/__manifest__.py]: The purchase is not required for this feature.
```suggestion
        "sale_management"
```

File: sale_require_po_doc/models/res_partner.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    customer_need_po = fields.Boolean(
+        string='Customer Requires PO', track_visibility="onchange",
```
Comment (dreispt) [inline sale_require_po_doc/models/res_partner.py:11]:
```suggestion
        string='Customer Requires PO', track_visibility="onchange",
        help="A PO number will be required on the"
        " Sales Order field for Customer Reference",
```

File: sale_require_po_doc/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    )
+    sale_doc = fields.Text(
+        string='Sales Require Documentation', track_visibility="onchange",
```
Comment (dreispt) [inline sale_require_po_doc/models/res_partner.py:16]:
```suggestion
        string='Sales Require Documentation', track_visibility="onchange",
        help="Details of the sales documentation required for this customer."
        " Sales documentation notes field will be required on Sales Orders"
        " This is complementary to requiring a PO number, and can be used"
        " with or without that setting.",
```

File: sale_require_po_doc/models/sale_order.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    customer_need_po = fields.Boolean(
+        related="partner_id.customer_need_po",
+        string='Customer Requires PO',
```
Comment (dreispt) [inline sale_require_po_doc/models/sale_order.py]:
```suggestion
        string='Customer Requires PO',
        read_only=True,
```

File: sale_require_po_doc/models/sale_order.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    )
+    sale_doc = fields.Text(
+        related="partner_id.sale_doc",
+        string='Sales Require Documentation',
```
Comment (dreispt) [inline sale_require_po_doc/models/sale_order.py]:
```suggestion
        string='Sales Require Documentation',
        read_only=True,
```

File: sale_require_po_doc/models/sale_order.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    )
+    sale_document_option = fields.Selection(
+        [('no_need', 'No documentation needed for this Sale'),
+         ('required', 'Documentation required obtained and archived')],
```
Comment (dreispt) [inline sale_require_po_doc/models/sale_order.py]:
```suggestion
         ('done', 'Documentation required obtained and archived')],
```

File: sale_require_po_doc/models/sale_order.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        for order in self:
+            if order.customer_need_po and not order.client_order_ref:
+                raise ValidationError(
+                    _("You can not confirm sale order without \
+                        Customer reference."))
+            if order.sale_doc and not order.sale_document_option:
+                raise ValidationError(
+                    _("You can not confirm sale order without \
+                        Sale Documentation."))
+        return res
```
Comment (dreispt) [inline sale_require_po_doc/models/sale_order.py]: Less resources spend if we do the validation upfront
```suggestion
        for order in self:
            if order.customer_need_po and not order.client_order_ref:
                raise ValidationError(
                    _("You can not confirm sale order without \
                        Customer reference."))
            if order.sale_doc and not order.sale_document_option:
                raise ValidationError(
                    _("You can not confirm sale order without \
                        Sale Documentation."))
        return super(SaleOrder, self).action_confirm()
```

File: sale_require_po_doc/views/sale_order_view.xml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    <record id="sales_po_sale_order_custom" model="ir.ui.view">
+        <field name="name">sale.order.sales.po.custom</field>
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_order_form"/>
+        <field name="arch" type="xml">
+            <field name="client_order_ref" position="after">
+                <field name="customer_need_po"/>
+                <field name="sale_doc"/>
+                <field attrs="{'invisible':[('sale_doc','=','')]}" name="sale_document_option"/>
+                <field attrs="{'invisible':[('sale_doc','=','')]}" name="sale_document_note"/>
```
Comment (dreispt) [inline sale_require_po_doc/views/sale_order_view.xml]: Fix, we also want to make it visible of sale_document_option is set:
```suggestion
                <field attrs="{'invisible':['&', ('sale_doc', '=', False), ('sale_document_option', '=', False)]}" 
                          name="sale_document_option"/>
                <field attrs="{'invisible':['&', ('sale_doc', '=', False), ('sale_document_option', '=', False)]}" 
                          name="sale_document_note"/>
```

**Review by dreispt — COMMENTED** [review]
One fix and severall improvements were suggested.
Good to merge as soon as they are incorporated.

**Review by dreispt — APPROVED** [review]
👍

**Pull request merged**